BlueDBOnDisk/src/main/java/io/bluedb/disk/segment/Segment.java

```diff
@@ -1,94 +1,201 @@
+	}
+
+	public boolean contains(BlueKey key) throws BlueDbException {
+		File file = getFileFor(key);
+		if (!file.exists()) {
+			return false;
+		}
+		Map<BlueKey, T> fileContents = fetchAsMap(file);
+		return fileContents.containsKey(key);
 	}
 
 	public void put(BlueKey key, T value) throws BlueDbException {
-		TreeMap<BlueKey, BlueEntity<T>> data = load();
-		BlueEntity<T> entity = new BlueEntity<T>(key, value);
-		data.put(key, entity);
-		Blutils.save(pathString, data, serializer);
+		File file = getFileFor(key);
+		Map<BlueKey, T> fileContents = fetchAsMap(file);
+		fileContents.put(key, value);
+		persist(file, fileContents);
 	}
 
 	public void delete(BlueKey key) throws BlueDbException {
-		TreeMap<BlueKey, BlueEntity<T>> data = load();
-		data.remove(key);
-		Blutils.save(pathString, data, serializer);
+		File file = getFileFor(key);
+		if (file.exists()) {
+			Map<BlueKey, T> fileContents = fetchAsMap(file);
+			fileContents.remove(key);
+			persist(file, fileContents);
+		}
 	}
 
-	public BlueEntity<T> read(BlueKey key) throws BlueDbException {
-		Collection<BlueEntity<T>> values = load().values();
-		for (BlueEntity<T> entity: values) {
-			if (entity.getKey().equals(key))
-				return entity;
-		}
-		return null;
+	public T get(BlueKey key) throws BlueDbException {
+		File file = getFileFor(key);
+		Map<BlueKey, T> fileContents = fetchAsMap(file);
+		return fileContents.get(key);
 	}
 
-	public List<BlueEntity<T>> read() throws BlueDbException {
-		List<BlueEntity<T>> results = new ArrayList<>(load().values());
+	public List<T> getAll() throws BlueDbException {
+		File[] filesInFolder = segmentPath.toFile().listFiles();
+		List<T> results = new ArrayList<>();
+		for (File file: filesInFolder) {
+			Map<BlueKey, T> fileContents = fetchAsMap(file);
+			results.addAll(fileContents.values());
+		}
 		return results;
 	}
 
-	public List<BlueEntity<T>> read(long minTime, long maxTime) throws BlueDbException {
+    public List<BlueEntity<T>> getRange(long minTime, long maxTime) throws BlueDbException {
+		File[] filesInFolder = segmentPath.toFile().listFiles();
 		List<BlueEntity<T>> results = new ArrayList<>();
-		for (BlueEntity<T>entity: load().values()) {
-			if (Blutils.meetsTimeConstraint(entity.getKey(), minTime, maxTime))
-				results.add(entity);
+		for (File file: filesInFolder) {
+			List<BlueEntity<T>> fileContents = fetch(file);
+			for (BlueEntity<T> entity: fileContents) {
+				BlueKey key = entity.getKey();
+				if (inTimeRange(minTime, maxTime, key)) {
+					results.add(entity);
+				}
+			}
 		}
 		return results;
 	}
 
+	protected File getFileFor(BlueKey key) {
+		return getPathFor(key).toFile();
+	}
+
+	protected Path getPath() {
+		return segmentPath;
+	}
+
+	private Path getPathFor(BlueKey key) {
+		String fileName = String.valueOf(key.getGroupingNumber());
+		return Paths.get(segmentPath.toString(), fileName);
+	}
+
+	private TreeMap<BlueKey, T> fetchAsMap(File file) throws BlueDbException {

```

Comment:
You have to iterate through every item in the list to convert it to a map and every item in the map to convert it back to a list. Can't we just iterate through the list one time to identify and update/delete the item(s) that we care about? That would save a lot of cycles and memory. I don't see what the map gives us that isn't pretty easy to do on our own.